### PR TITLE
Redesign HealthStatus (again)

### PR DESCRIPTION
**Pull request by rynowak**
This change brings back the ability to return Healthy/Degraded/Unhealthy
in a HealthCheckResult. We tried making this pass/fail in 2.2.0-preview3
and folks writing health checks for their own use pointed out (rightly
so) that it was too limited.

It's still possible for the app developer to configure the failure
status of a health check, but it requires the health check author to
cooperate.

I also got rid of HealthStatus.Failed since it raises more questions
than it answers. It's really not clear that it's valuable for a health
check for behave different when throwing an unhandled exception.

We would still recommend that a health check library handle exceptions
that they know about and return `context.Registration.FailureStatus`.

### CODE REVIEW

**Comment by unaizorrilla** [comment]
Hi @rynowak 

I try to adapt all HealthChecks / UI  to this new HealthCheckResult status soon!

File: samples/HealthChecksSample/DbConnectionHealthCheck.cs [hunk 1]
```diff
@@ -51,11 +51,11 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
                 }
                 catch (DbException ex)
                 {
-                    return HealthCheckResult.Failed(exception: ex);
+                    return new HealthCheckResult(status: context.Registration.FailureStatus, exception: ex);
```
Comment (rynowak) [inline samples/HealthChecksSample/DbConnectionHealthCheck.cs]: @unaizorrilla - I would expect your code to look a lot like this. This pattern allows the app developer to configure the status.

Comment (unaizorrilla) [reply samples/HealthChecksSample/DbConnectionHealthCheck.cs]: Hi @rynowak ,
 
Yes, when the new bits will be available on MyGet I perfom the changes on HealthChecks

File: src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs [hunk 2]
```diff
@@ -49,29 +49,43 @@ public HealthCheckResult(bool result, string description, Exception exception, I
         /// is considered to have passed health validation. A <c>false</c> value will be mapped to the configured 
         /// <see cref="HealthStatus"/> by  the health check system.
         /// </summary>
-        public bool Result { get; }
+        public HealthStatus Status { get; }
 
         /// <summary>
         /// Creates a <see cref="HealthCheckResult"/> representing a passing component.
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs]: > a ~~passing~~ healthy component

File: src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs [hunk 3]
```diff
@@ -49,29 +49,43 @@ public HealthCheckResult(bool result, string description, Exception exception, I
...
         /// </summary>
         /// <returns>A <see cref="HealthCheckResult"/> representing a passing component.</returns>
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs]: passing -> healthy

File: src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs [hunk 4]
```diff
@@ -49,29 +49,43 @@ public HealthCheckResult(bool result, string description, Exception exception, I
...
+        /// <param name="description">A human-readable description of the status of the component that was checked. Optional.</param>
+        /// <param name="exception">An <see cref="Exception"/> representing the exception that was thrown when checking for status. Optional.</param>
+        /// <param name="data">Additional key-value pairs describing the health of the component. Optional.</param>
+        /// <returns>A <see cref="HealthCheckResult"/> representing a degraged component.</returns>
+        public static HealthCheckResult Degraded(string description = null, Exception exception = null, IReadOnlyDictionary<string, object> data = null)
+        {
+            return new HealthCheckResult(status: HealthStatus.Degraded, description, exception: null, data);
+        }
+
+
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs]: Extra line

File: src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs [hunk 5]
```diff
@@ -49,29 +49,43 @@ public HealthCheckResult(bool result, string description, Exception exception, I
...
         /// <summary>
         /// Creates a <see cref="HealthCheckResult"/> representing an failing component.
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs]: unhealthy

File: src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthStatus.cs [hunk 6]
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.Extensions.Diagnostics.HealthChecks
     /// </summary>
     /// <remarks>
     /// <para>
-    /// A health status is derived the pass/fail result of an <see cref="IHealthCheck"/> (<see cref="HealthCheckResult.Result"/>)
+    /// A health status is derived the pass/fail result of an <see cref="IHealthCheck"/> (<see cref="HealthCheckResult.Status"/>)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthStatus.cs]: is derived **from** the  pass/fail

**Comment by unaizorrilla** [comment]
Hi @rynowak 

Committed HealthStatus re-design on AspNetcore.Diagnostics.HealthChecks repo. 

https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/commit/051947206d4837608de92f84c02979515868a5ce  

Versions 2.2.0-rtm-35613 on MyGet right now.

**Pull request merged**